metricbeat/module/windows/perfmon/reader.go

```diff
@@ -0,0 +1,172 @@
+						logp.Namespace("perfmon"), "query", counter.Query)
+					continue
+				}
+			} else {
+				query.Close()
+				return nil, errors.Wrapf(err, `failed to expand counter (query="%v")`, counter.Query)
+			}
+		}
+		if childQueries == nil || len(childQueries) == 0 {
+			return nil, errors.Wrapf(err, `failed to expand counter (query="%v")`, counter.Query)
+		}
+		for _, v := range childQueries {
+			if err := query.AddCounter(v, counter, len(childQueries) > 1); err != nil {
+				query.Close()
+				return nil, errors.Wrapf(err, `failed to add counter (query="%v")`, counter.Query)
+			}
+
+			r.instanceLabel[v] = counter.InstanceLabel
+			r.measurement[v] = counter.MeasurementLabel
+		}
+	}
+
+	return r, nil
+}
+
+// Read executes a query and returns those values in an event.
+func (r *Reader) Read() ([]mb.Event, error) {
+	if err := r.query.CollectData(); err != nil {
+		return nil, errors.Wrap(err, "failed querying counter values")
+	}
+
+	// Get the values.
+	values, err := r.query.GetFormattedCounterValues()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed formatting counter values")
+	}
+
+	eventMap := make(map[string]*mb.Event)
+
+	for counterPath, values := range values {
+		for ind, val := range values {
+			if val.Err != nil && !r.executed {
+				r.log.Debugw("Ignoring the first measurement because the data isn't ready",
+					"error", val.Err, logp.Namespace("perfmon"), "query", counterPath)
+				continue
+			}
+
+			var eventKey string
+			if r.groupMeasurements && val.Err == nil {
+				// Send measurements with the same instance label as part of the same event
+				eventKey = val.Instance
+			} else {
+				// Send every measurement as an individual event
+				// If a counter contains an error, it will always be sent as an individual event
+				eventKey = counterPath + strconv.Itoa(ind)
+			}
+
+			// Create a new event if the key doesn't exist in the map
+			if _, ok := eventMap[eventKey]; !ok {
+				eventMap[eventKey] = &mb.Event{
+					MetricSetFields: common.MapStr{},
+					Error:           errors.Wrapf(val.Err, "failed on query=%v", counterPath),
+				}
+				if val.Instance != "" {
+					//will ignore instance counter
+					if ok, match := matchesParentProcess(val.Instance); ok {
+						eventMap[eventKey].MetricSetFields.Put(r.instanceLabel[counterPath], match)
+					} else {
+						eventMap[eventKey].MetricSetFields.Put(r.instanceLabel[counterPath], val.Instance)
+					}
+				}
+			}
+			event := eventMap[eventKey]
+			if val.Measurement != nil {
+				event.MetricSetFields.Put(r.measurement[counterPath], val.Measurement)
+			} else {
+				event.MetricSetFields.Put(r.measurement[counterPath], 0)
+			}
+		}
+	}
+
+	// Write the values into the map.
+	events := make([]mb.Event, 0, len(eventMap))
+	for _, val := range eventMap {
+		events = append(events, *val)
+	}
+
+	r.executed = true
+	return events, nil
+}
+
+// CloseQuery will close the PDH query.
+func (r *Reader) CloseQuery() error {
```

Comment:
Just call this method `Close`, so it doesn't expose implementation details and can be used if more things need to be closed in a Reader in the future.
```suggestion
func (r *Reader) Close() error {
```